pkg/splunk/enterprise/util.go

```diff
@@ -882,3 +866,85 @@ func setupAppInitContainers(client splcommon.ControllerClient, cr splcommon.Meta
+
+	appInfoStatus.LastAppInfoCheckTime = currentEpoch
+}
+
+// HasAppRepoCheckTimerExpired checks if the polling interval has expired
+func HasAppRepoCheckTimerExpired(appFrameworkRef *enterprisev1.AppFrameworkSpec, appInfoContext *enterprisev1.AppDeploymentContext) bool {
+	scopedLog := log.WithName("hasAppRepoCheckTimerExpired")
+	currentEpoch := time.Now().Unix()
+	scopedLog.Info("Checking if the app repo polling interval has expired", "LastAppInfoCheckTime", strconv.FormatInt(appInfoContext.LastAppInfoCheckTime, 10), "current epoch time", strconv.FormatInt(currentEpoch, 10))
+	return appInfoContext.LastAppInfoCheckTime+int64(appFrameworkRef.AppsRepoPollInterval) <= currentEpoch
+}
+
+// GetNextRequeueTime gets the next reconcile requeue time based on the appRepoPollInterval.
+// There can be some time elapsed between when we frst set lastAppInfoCheckTime and when the CR is in Ready state.
+// Hence we need to subtracct the delta time elapsed from the actual polling interval,
+// so that the next reconile would happen at the right time.
+func GetNextRequeueTime(appRepoPollInterval, lastCheckTime int64) time.Duration {
+	scopedLog := log.WithName("GetNextRequeueTime")
+	currentEpoch := time.Now().Unix()
+
+	var nextRequeueTimeInSec int64
+	nextRequeueTimeInSec = appRepoPollInterval - (currentEpoch - lastCheckTime)
+
+	scopedLog.Info("Getting next requeue time", "LastAppInfoCheckTime", lastCheckTime, "Current Epoch time", currentEpoch, "nextRequeueTimeInSec", nextRequeueTimeInSec)
+
+	return time.Second * (time.Duration(nextRequeueTimeInSec))
+}
+
+// initAndCheckAppInfoStatus initializes the S3Clients and checks the status of apps on remote storage.
+func initAndCheckAppInfoStatus(client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkConf *enterprisev1.AppFrameworkSpec, appStatusContext *enterprisev1.AppDeploymentContext) error {
+	scopedLog := log.WithName("initAndCheckAppInfoStatus").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+
+	var err error
+	// Register the S3 clients specific to providers if not done already
+	// This is done to prevent the null pointer dereference in case when
+	// operator crashes and comes back up and the status of app context was updated
+	// to match the spec in the previous run.
+	initAppFrameWorkContext(appFrameworkConf, appStatusContext)
+
+	//check if the apps need to be downloaded from remote storage
+	if HasAppRepoCheckTimerExpired(appFrameworkConf, appStatusContext) || !reflect.DeepEqual(appStatusContext.AppFrameworkConfig, *appFrameworkConf) {
+		var sourceToAppsList map[string]splclient.S3Response
+
+		scopedLog.Info("Checking status of apps on remote storage...")
+
+		sourceToAppsList, err = GetAppListFromS3Bucket(client, cr, appFrameworkConf)
+		if len(sourceToAppsList) != len(appFrameworkConf.AppSources) {
+			scopedLog.Error(err, "Unable to get apps list, will retry in next reconcile...")
+		} else {
```

Comment:
I think this is something that we need to strengthen how we want to handle, particularly for phase 3. With the current state machine logic, we mark any missing appSrc as deleted in the operator state. The only downsize for phase 2 is, once that appSrc is up next time,  we reset the Pod, and try to install the app again. If this behavior is fine, we can go ahead with a ToDo item for phase-3.

Reply:
Yeah probably fine for now.

Reply:
Ok, probably, we can have a ToDo comment for now? Alternatively, a separate task is a better idea? as we have to do some amount of code changes for handling this case?

Reply:
Ok I will add a TODO comment for now, we can take it up later in phase 3